pymodbus/client/async/asyncio/__init__.py

```diff
@@ -0,0 +1,777 @@
+"""
+Asynchronous framework adapter for asyncio.
+"""
+import socket
+import asyncio
+import functools
+from pymodbus.exceptions import ConnectionException
+from pymodbus.client.async import AsyncModbusClientMixin
+from pymodbus.compat import byte2int
+import logging
+
+_logger = logging.getLogger(__name__)
+
+DGRAM_TYPE = socket.SocketKind.SOCK_DGRAM
+
+
+class BaseModbusAsyncClientProtocol(AsyncModbusClientMixin):
+    """
+    Asyncio specific implementation of asynchronous modbus client protocol.
+    """
+
+    #: Factory that created this instance.
+    factory = None
+    transport = None
+
+    def connection_made(self, transport):
+        """
+        Called when a connection is made.
+
+        The transport argument is the transport representing the connection.
+        :param transport:
+        :return:
+        """
+        self.transport = transport
+        self._connectionMade()
+
+        if self.factory:
+            self.factory.protocol_made_connection(self)
+
+    def connection_lost(self, reason):
+        """
+        Called when the connection is lost or closed.
+
+        The argument is either an exception object or None
+        :param reason:
+        :return:
+        """
+        self.transport = None
+        self._connectionLost(reason)
+
+        if self.factory:
+            self.factory.protocol_lost_connection(self)
+
+    def data_received(self, data):
+        """
+        Called when some data is received.
+        data is a non-empty bytes object containing the incoming data.
+        :param data:
+        :return:
+        """
+        self._dataReceived(data)
+
+    def create_future(self):
```

Comment:
How using this helper is better than using normal asyncio interface? ;)

Reply:
@pax0r didn't get your question, could you elaborate?

Reply:
I am not sure if it's a good idea to create helper function which consists only of single call to some standard lib element. I believe it may be better to just directly use `asyncio.Future()`. But it's a nit, you can ignore it ;)